### PR TITLE
Add http testing framework

### PR DESCRIPTION
**Pull request by ajbt200128**
PR checklist:

- [x] Purpose of the code is [evident to future readers](https://semgrep.dev/docs/contributing/contributing-code/#explaining-code)
- [x] Tests included or PR comment includes a reproducible test plan
- [ ] Documentation is up-to-date
- [ ] A changelog entry was [added to changelog.d](https://semgrep.dev/docs/contributing/contributing-code/#adding-a-changelog-entry) for any user-facing change
- [ ] Change has no security implications (otherwise, ping security team)

If you're unsure about any of this, please see:

- [Contribution guidelines](https://semgrep.dev/docs/contributing/contributing-code)!
- [One of the more specific guides located here](https://semgrep.dev/docs/contributing/contributing/)

### CODE REVIEW

File: libs/networking/Testing_client.mli [hunk 1]
```diff
@@ -0,0 +1,14 @@
+type test_response = { response : Cohttp_lwt.Response.t; body_path : string }
```
Comment (aryx) [inline libs/networking/Testing_client.mli:31]: maybe you can add some explanations in the .mli on how to use this library. Maybe the code of a full test using
those mocks.

File: libs/networking/http_helpers.ml [hunk 2]
```diff
@@ -1,17 +1,26 @@
...
 (* This way we can use the async methods in the language server,
    and other places too
 *)
 
+(*****************************************************************************)
+(* Client *)
+(*****************************************************************************)
+
+(* Create a client reference so we can swap it out with a testing version *)
+let client_ref = ref (module Client : Cohttp_lwt.S.Client)
```
Comment (aryx) [inline libs/networking/http_helpers.ml]: first class modules!

**Review by aryx — COMMENTED** [review]
Nice! Just need an example now.

File: libs/networking/http_helpers.ml (hunk 2 above)
Comment (akuhlens) [reply libs/networking/http_helpers.ml]: I didn't know you could do that!

Comment (aryx) [reply libs/networking/http_helpers.ml]: https://dev.realworldocaml.org/first-class-modules.html#:~:text=First%2Dclass%20modules%20are%20ordinary,language%20to%20use%20them%20effectively.
But we should avoid using such advanced features. We should use them only as a last resort (and this look like a good use case given the way the Cohttp library is written).

File: libs/networking/Testing_client.mli (hunk 1 above)
Comment (ajbt200128) [reply libs/networking/Testing_client.mli:31]: Yea, was just a POC to show andre, should be updated now

File: libs/networking/http_helpers.ml (hunk 2 above)
Comment (ajbt200128) [reply libs/networking/http_helpers.ml]: Yea, it was either this or functors and that seemed too heavy handed.

**Comment by ajbt200128** [comment]
Also added a test bytecode target, and an action to run tests w/ earlybird. And added a basic LS test

File: src/osemgrep/language_server/Unit_LS.ml [hunk 3]
```diff
@@ -224,5 +224,25 @@ let processed_run () =
   in
   pack_tests "Processed Run" tests
 
+let session_rules () =
+  let with_ci_client =
+    let make_fn req body =
+      ignore body;
+      Testing_client.check_method req "GET";
+      Lwt.return Testing_client.(basic_response "./tests/ls/ci/response.json")
```
Comment (akuhlens) [inline src/osemgrep/language_server/Unit_LS.ml:232]: Could you ensure that there is some easy way to check the message against a file here? My concern is that Request objects from cohttp are pretty opaque, just want to ensure that there isn't more code needed for this, or if we need to introduce a non-opaque request similar to `Rule.request`

File: src/tests/dune [hunk 4]
```diff
@@ -25,6 +25,7 @@
     osemgrep_targeting
     osemgrep_language_server
   )
+ (modes native byte)
```
Comment (aryx) [inline src/tests/dune]: maybe add a comment about byte, which is it allows to produce a test.bc and use
earlybird on it

File: Makefile [hunk 5]
```diff
@@ -188,6 +188,12 @@ core-test:
 	./_build/default/src/tests/test.exe --show-errors --help 2>&1 >/dev/null
 	./scripts/run-core-test
 
+.PHONY: core-test-bc
+core-test-bc:
```
Comment (aryx) [inline Makefile]: you can remove the core- prefix here

File: .vscode/tasks.json [hunk 6]
```diff
@@ -26,6 +26,14 @@
         "kind": "build"
       }
     },
+    {
+      "label": "Build Test Bytecode",
+      "type": "shell",
+      "command": "make core-test-bc",
```
Comment (aryx) [inline .vscode/tasks.json]: make test-bc should be enough

**Review by akuhlens — COMMENTED** [review]
This looks great! Thank you. Follow up on the comment that I made. I am okay with the majority of the work for that actually being in a seperate PR if need be, as long as you are confident the interface doesn't need to change.

File: libs/networking/Testing_client.ml [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+ * special exception on linking described in file LICENSE.
+ *
+ * This library is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the file
+ * LICENSE for more details.
+ *)
+
+(* Commentary *)
+(* This is a testing client to mock out http_helpers.ml so we
```
Comment (aryx) [inline libs/networking/Testing_client.ml]: better to put that in the .mli.
In theory people should just read the interface file (.mli) to understand how to use your library.
They can look at the .ml to understand the implementation, get possibly comments about design decisions.

**Review by aryx — APPROVED** [review]

File: src/osemgrep/language_server/Unit_LS.ml (hunk 3 above)
Comment (ajbt200128) [reply src/osemgrep/language_server/Unit_LS.ml:232]: @akuhlens there's a `check_body` helper function. If you check the header of Testing_client, there should be an example there

File: libs/networking/Testing_client.ml [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+           (Cohttp.Header.to_string (Cohttp.Request.headers req)))
+  | Some actual_header ->
+      Alcotest.(check string) "header" header_val actual_header
+
+(*****************************************************************************)
+(* Entrypoint *)
+(*****************************************************************************)
+
+let with_testing_client make_fn test_fn () =
+  let prev_client = !Http_helpers.client_ref in
```
Comment (aryx) [inline libs/networking/Testing_client.ml:110]: You should use Common.save_excursion (it's a function I wrote inspired by the same function in emacs lisp).

Comment (aryx) [reply libs/networking/Testing_client.ml:110]: It's more tolerant to exn that could be thrown in test_fn ()

File: src/tests/dune [hunk 9]
```diff
@@ -25,6 +25,7 @@
     osemgrep_targeting
     osemgrep_language_server
   )
+ (modes native) ; bytecode here for debugging
```
Comment (aryx) [inline src/tests/dune:28]: so why did you remove the bytecode from the modes?

File: libs/networking/Testing_client.ml (hunk 8 above)
Comment (ajbt200128) [reply libs/networking/Testing_client.ml:110]: I didn't know about that, very cool

File: src/tests/dune (hunk 9 above)
Comment (ajbt200128) [reply src/tests/dune:28]: Forgot to remove the comment. Going to revisit this today in a separate PR (wanted to get this merged for andre)

**Pull request merged**